router/router.go

```diff
@@ -0,0 +1,340 @@
+	cfgfile     string
+	aquaServer  string
+	outputs     map[string]outputs.Output
+	inputRoutes map[string]*routes.InputRoute
+	templates   map[string]data.Inpteval
+}
+
+var (
+	errNoOutputs  = errors.New("there aren't started outputs")
+	initCtx       sync.Once
+	routerCtx     *Router
+	baseForTicker = time.Hour
+
+	osStat = os.Stat
+
+	requireAuthorization = map[string]bool{
+		"servicenow": true,
+		"jira":       true,
+	}
+)
+
+func Instance() *Router {
+	initCtx.Do(func() {
+		routerCtx = &Router{
+			mutexScan:   sync.Mutex{},
+			quit:        make(chan struct{}),
+			queue:       make(chan []byte, 1000),
+			outputs:     make(map[string]outputs.Output),
+			inputRoutes: make(map[string]*routes.InputRoute),
+			templates:   make(map[string]data.Inpteval),
+			stopTicker:  make(chan struct{}),
+		}
+	})
+	return routerCtx
+}
+func (ctx *Router) ReloadConfig() {
+	ctx.Terminate()
+	ctx.Start(ctx.cfgfile)
+}
+
+func (ctx *Router) Start(cfgfile string) error {
+	log.Printf("Starting Router....")
+
+	ctx.cfgfile = cfgfile
+	ctx.outputs = map[string]outputs.Output{}
+	ctx.inputRoutes = map[string]*routes.InputRoute{}
+	ctx.templates = map[string]data.Inpteval{}
+	ctx.ticker = nil
+
+	err := ctx.load()
+	if err != nil {
+		return err
+	}
+	go ctx.listen()
+	return nil
+}
+
+func (ctx *Router) Terminate() {
+	log.Printf("Terminating Router....")
+
+	for _, pl := range ctx.outputs {
+		pl.Terminate()
+	}
+	log.Printf("Outputs terminated")
+
+	for _, route := range ctx.inputRoutes {
+		route.StopScheduler()
+	}
+	log.Printf("Route schedulers stopped")
+
+	ctx.quit <- struct{}{}
+	log.Printf("quit notified")
+	if ctx.ticker != nil {
+		ctx.stopTicker <- struct{}{}
+		log.Printf("stopTicker notified")
+	}
+
+}
+
+func (ctx *Router) Send(data []byte) {
+	ctx.queue <- data
+}
+
+func (ctx *Router) initTemplate(template *Template) error {
+	log.Printf("Configuring template %s \n", template.Name)
+
+	if template.LegacyScanRenderer != "" {
+		inpteval, err := formatting.BuildLegacyScnEvaluator(template.LegacyScanRenderer)
+		if err != nil {
+			return err
+		}
+		ctx.templates[template.Name] = inpteval
+		log.Printf("Configured with legacy renderer %s \n", template.LegacyScanRenderer)
+	}
+
+	if template.RegoPackage != "" {
+		inpteval, err := regoservice.BuildBundledRegoEvaluator(template.RegoPackage)
+		if err != nil {
+			return err
+		}
+		ctx.templates[template.Name] = inpteval
+		log.Printf("Configured with Rego package %s\n", template.RegoPackage)
+	}
+	if template.Url != "" {
+		log.Printf("Configured with url: %s\n", template.Url)
+
+		r, err := http.NewRequest("GET", template.Url, nil)
+		if err != nil {
+			return err
+		}
+		httpClient := getHttpClient()
+		resp, err := httpClient.Do(r)
+		if err != nil {
+			return err
+		}
+
+		if resp.StatusCode > 399 {
+			return errors.New(fmt.Sprintf("can not connect to %s, response status is %d", template.Url, resp.StatusCode))
+		}
+
+		b, err := ioutil.ReadAll(resp.Body)
+		defer resp.Body.Close()
```

Comment:
I'd transfer this `defer` call after the check of error.

Reply:
done